app/uk/gov/hmrc/gform/obligation/ObligationService.scala

```diff
@@ -17,30 +17,123 @@
+    authService: AuthService,
+    retrievals: MaterialisedRetrievals,
+    form: Form,
+    hmrcTaxPeriodIdentifiers: NonEmptyList[HmrcTaxPeriod])(implicit hc: HeaderCarrier, ec: ExecutionContext) = {
+
+    for {
+      listOfIdNumbers <- {Future.traverse(hmrcTaxPeriodIdentifiers.toList)(
+                          i =>
+                            authService.evaluateSubmissionReference(
+                              i.idNumber,
+                              retrievals,
+                              formTemplate,
+                              FormDataHelpers.formDataMap(form.formData),
+                              form.envelopeId))}
+      output <- gformConnector
+                 .getAllTaxPeriods(
+                   NonEmptyList.fromList(
+                     listOfIdNumbers
+                       .zip(hmrcTaxPeriodIdentifiers.toList)
+                       .map(a => HmrcTaxPeriod(a._2.idType, TextExpression(Constant(a._1)), a._2.regimeType))) match {
+                     case Some(x) => x
+                     case None => NonEmptyList(HmrcTaxPeriod(IdType("NONE"), TextExpression(Constant("NONE")), RegimeType("NONE")), List())
+                   })
+                 .map(i =>
+                   i.flatMap(j => j.obligation.obligations.flatMap(h => makeAllInfoList(j.id, h.obligationDetails))))
+    } yield output
+  }
+
+  def makeAllInfoList(id: HmrcTaxPeriod, obligation: NonEmptyList[ObligationDetail]) =
+    obligation.map(i =>
+      TaxPeriodInformation(id, i.inboundCorrespondenceFromDate, i.inboundCorrespondenceToDate, i.periodKey))
+
+  def lookupIfPossible(
+    form: Form,
+    formTemplate: FormTemplate,
+    authService: AuthService,
+    retrievals: MaterialisedRetrievals)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Form] = {
+
     val hmrcTaxPeriodIdentifiers = formTemplate.expandFormTemplateFull.allFCs.collect {
       case IsHmrcTaxPeriod(el) => el
     }
-    val futureListOfTaxPeriodDes =
-      gformConnector
-        .getAllTaxPeriods(hmrcTaxPeriodIdentifiers)
-        .map(i => i.flatMap(j => j.obligation.obligations.flatMap(h => makeMap(j.id, h.obligationDetails))))
-    futureListOfTaxPeriodDes.map(i => i.toMap)
+    val hmrcTaxPeriodIdentifiersNonEmpty: NonEmptyList[HmrcTaxPeriod] = NonEmptyList.fromList(hmrcTaxPeriodIdentifiers) match {
+      case Some(x) => x
+      case _ => NonEmptyList(HmrcTaxPeriod(IdType("NONE"), TextExpression(Constant("NONE")), RegimeType("NONE")), List())
+    }
+
+    def ifStatement(condition: Boolean) =
+      if (condition) {
+        val newObligations = lookupObligationsMultiple(formTemplate, authService, retrievals, form, hmrcTaxPeriodIdentifiersNonEmpty)
+        newObligations.map(i => form.copy(obligations = RetrievedObligations(i)))
+      } else {
+        Future.successful(form)
+      }
+
+    hmrcTaxPeriodIdentifiers match {
+      case a :: _ =>
+        for {
+          currentIdNumber <- authService.evaluateSubmissionReference(
```

Comment:
(Continuation form https://github.com/hmrc/gform-frontend/pull/643/files#r260219771)

All this call should never happen, nor should `eval` be called here directly (the same holds for line 48 and 109). 

What should happen instead is that `eval` should evaluate `HmrcTaxPeriod` component's `idNumber` expression (when called as part of `ProcessDataService.getProcessData`) and it should be given back to this service as part of `FormDataRecalculated` class instance. 

If we cannot get hold of `FormDataRecalculated` instance, something is wrong in whole design and bigger refactor/rethinking will be needed.